stl/inc/compare

```diff
@@ -470,6 +474,79 @@ struct common_comparison_category {
 template <class... _Types>
 using common_comparison_category_t = typename common_comparison_category<_Types...>::type;
 
+#if defined(__cpp_impl_three_way_comparison) && defined(__cpp_lib_concepts)
+// clang-format off
+template <class _Ty1, class _Ty2>
+    requires _Is_comparison_category<_Ty1> && _Is_comparison_category<_Ty2>
+struct _Common_type2<_Ty1, _Ty2, _Ty1, _Ty2> : common_comparison_category<_Ty1, _Ty2> {};
+
+template <class _Ty, class _Cat>
+concept _Compares_as = same_as<common_comparison_category_t<_Ty, _Cat>, _Cat>;
+
+template <class _Ty, class _Cat = partial_ordering>
+concept three_way_comparable = _Weakly_equality_comparable_with<_Ty, _Ty> && _Partially_ordered_with<_Ty, _Ty>
+    && requires(const remove_reference_t<_Ty>& __a, const remove_reference_t<_Ty>& __b) {
+        { __a <=> __b } -> _Compares_as<_Cat>;
+    };
+
+template <class _Ty1, class _Ty2, class _Cat = partial_ordering>
+concept three_way_comparable_with = three_way_comparable<_Ty1, _Cat> && three_way_comparable<_Ty2, _Cat>
+    && common_reference_with<const remove_reference_t<_Ty1>&, const remove_reference_t<_Ty2>&>
+    && three_way_comparable<common_reference_t<const remove_reference_t<_Ty1>&, const remove_reference_t<_Ty2>&>, _Cat>
+    && _Weakly_equality_comparable_with<_Ty1, _Ty2> && _Partially_ordered_with<_Ty1, _Ty2>
+    && requires(const remove_reference_t<_Ty1>& __t, const remove_reference_t<_Ty2>& __u) {
+        { __t <=> __u } -> _Compares_as<_Cat>;
+        { __u <=> __t } -> _Compares_as<_Cat>;
+    };
+
+template <class _Ty1, class _Ty2 = _Ty1>
+using compare_three_way_result_t =
+    decltype(_STD declval<const remove_reference_t<_Ty1>>() <=> _STD declval<const remove_reference_t<_Ty2>>());

```

Comment:
Are const rvalues intentionally being formed here?

Reply:
Sorry, that's a small mistake from my side. These should be const lvalue references.